# Выбор семейства распределений в наивном байесе
В этом задании вы сможете немного расслабиться после реализации случайного леса и градиентного бустинга по мотивам материалов прошлой недели. Все, что нужно будет делать — запускать методы из sklearn. Вам предлагается выяснить, какое распределение лучше использовать в наивном байесовском классификаторе в зависимости от вида признаков.

Загрузите датасеты digits и breast_cancer из sklearn.datasets. Выведите несколько строчек из обучающих выборок и посмотрите на признаки. С помощью sklearn.cross_validation.cross_val_score c настройками по умолчанию и вызова метода mean() у возвращаемого этой функцией numpy.ndarray, сравните качество работы наивных байесовских классификаторов на этих двух датасетах. Для сравнения предлагается использовать BernoulliNB, MultinomialNB и GaussianNB. Насколько полученные результаты согласуются с рекомендациями из лекций?

Два датасета, конечно, еще не повод делать далеко идущие выводы, но при желании вы можете продолжить исследование на других выборках (например, из UCI репозитория). 

In [1]:
from sklearn import datasets,  naive_bayes
from sklearn.model_selection import cross_val_score
digits = datasets.load_digits()
cancer = datasets.load_breast_cancer()
# print(breast_cancer['DESCR'])
print(cancer.target_names)


['malignant' 'benign']


In [2]:
print('### digits')
print ('Bernoulli: ', cross_val_score(naive_bayes.BernoulliNB(), digits['data'], digits['target']).mean())
print ('Multinomial: ', cross_val_score(naive_bayes.MultinomialNB(), digits['data'], digits['target']).mean())
print ('Gaussian: ', cross_val_score(naive_bayes.GaussianNB(), digits['data'], digits['target']).mean())

### digits
Bernoulli:  0.8241736304549674
Multinomial:  0.8703497369235531
Gaussian:  0.8069281956050759


In [3]:
print('### breast_cancer')
print ('Bernoulli: ', cross_val_score(naive_bayes.BernoulliNB(), cancer['data'], cancer['target']).mean())
print ('Multinomial: ', cross_val_score(naive_bayes.MultinomialNB(), cancer['data'], cancer['target']).mean())
print ('Gaussian: ', cross_val_score(naive_bayes.GaussianNB(), cancer['data'], cancer['target']).mean())

### breast_cancer
Bernoulli:  0.6274181027790716
Multinomial:  0.8963204471355379
Gaussian:  0.9385188635305075


Для сдачи задания, ответьте на приведенные ниже вопросы.
#### Вопрос 1

Каким получилось максимальное качество классификации на датасете breast_cancer?

0.9385188635305075

#### Вопрос 2

Каким получилось максимальное качество классификации на датасете digits?

0.8703497369235531

#### Вопрос 3

Выберите верные утверждения и запишите их номера через пробел (в порядке возрастания номера):

1) На вещественных признаках лучше всего сработал наивный байесовский классификатор с распределением Бернулли - Нет

2) На вещественных признаках лучше всего сработал наивный байесовский классификатор с мультиномиальным распределением - Нет

3) Мультиномиальное распределение лучше показало себя на выборке с целыми неотрицательными значениями признаков - Да

4) На вещественных признаках лучше всего сработало нормальное распределение - Да

In [25]:
import numpy as np
np.sum([digits['data'] % 1])

0.0